include/camp/device.hpp

```diff
@@ -0,0 +1,311 @@
+    }
+    CudaEvent get_event() { 
+      return CudaEvent(get_stream());
+    }
+    Event get_event_erased() {
+      return Event{CudaEvent(get_stream())};
+    }
+    void wait() { cudaStreamSynchronize(stream); }
+    void wait_on(Event *e) { e->wait(); }
+    template <typename T>
+    T *allocate(size_t size)
+    {
+      T *ret = nullptr;
+      cudaMallocManaged(&ret, sizeof(T) * size);
+      return ret;
+    }
+    void *calloc(size_t size)
+    {
+      void *p = allocate<char>(size);
+      this->memset(p, 0, size);
+      return p;
+    }
+    void free(void *p) { cudaFree(p); }
+    void memcpy(void *dst, const void *src, size_t size)
+    {
+      cudaMemcpyAsync(dst, src, size, cudaMemcpyDefault, stream);
+    }
+    void memset(void *p, int val, size_t size)
+    {
+      cudaMemsetAsync(p, val, size, stream);
+    }
+
+    cudaStream_t get_stream() { return stream; }
+
+  private:
+    cudaStream_t stream;
+  };
+
+  class Host
+  {
+  public:
+    Host(int device = 0, int group = -1) {}
+
+    // Methods
+    Platform get_platform() { return Platform::host; }
+    Host &get_default()
+    {
+      static Host h;
+      return h;
+    }
+    HostEvent get_event() { return HostEvent(); }
+    Event get_event_erased() {
+      Event e{HostEvent()};
+      return e;
+    }
+    void wait() {} // nothing to wait for, sequential/simd host is always synchronous
+    void wait_on(Event *e) { }
+    // Memory
+    template <typename T>
+    T *allocate(size_t n)
+    {
+      return (T*)malloc(sizeof(T) * n);
+    }
+    void *calloc(size_t size)
+    {
+      void *p = allocate<char>(size);
+      this->memset(p, 0, size);
+      return p;
+    }
+    void free(void *p) { free(p); }
+    void memcpy(void *dst, const void *src, size_t size)
+    {
+      memcpy(dst, src, size);
+    }
+    void memset(void *p, int val, size_t size) { std::memset(p, val, size); }
+  };
+
+  class Context
+  {
+    public:
+      Context(){}
+      template<typename T>
+      Context(T&& value){ m_value.reset(new ContextModel<T>(value));}
+      template<typename T>
+      T* get_device() { 
+	auto result = dynamic_cast<ContextModel<T>*>(m_value.get()); 
+	if (result ==nullptr)
+	{
+	  std::runtime_error("Incompatible Context type get cast.");
+	}
+	return result->get_device();
+      }
+      Platform get_platform() { return m_value->get_platform(); }
+      template <typename T>
+      T *allocate(size_t size)
+      {
+	return (T *)m_value->calloc(size * sizeof(T));
+      }
+      void *calloc(size_t size) { return m_value->calloc(size); }
+      void free(void *p) { m_value->free(p); }
+      void memcpy(void *dst, const void *src, size_t size)
+      {
+	m_value->memcpy(dst, src, size);
+      }
+      void memset(void *p, int val, size_t size) { m_value->memset(p, val, size); }
+      Event get_event() { return m_value->get_event(); }
+      void wait_on(Event *e) { m_value->wait_on(e); }
+
+    private:
+      class ContextConcept {
```

Comment:
I'd prefer we call this ContextBase or ContextInterface or something like that.  Concept implies it as a checked concept for a class of types, and I actually want to add a context concept we can use for such checks to this code at some point.